### ISSUE TITLE
Issue with watching a deleted, then recreated directory

### ISSUE THREAD

**Issue by rottmann** [paulmillr/chokidar#917]
Problem exists before  #379

Create a dir, delete the dir, recreate the dir and add a file in the dir.
No event is fired.

**Versions (please complete the following information):**
 - Chokidar version 3.2.2
 - Node version 10.16.3
 - OS version: Mint 19.2

**To Reproduce:**

Create a dir, delete the dir, recreate the dir and add a file in the dir.

**Comment by apm963** [comment]
I noticed this too. I was able to get around it, but I had to build non-ideal logic to get it working. The basic steps I took were as follow:

1. Perform as `fs.access` check on each of the watched dirs to get the current status of whether they exist or not
2. If the dir doesn't exist, remove it (`watcher.unwatch`) from the watch list (`watcher.getWatched()`) if it is still being watched; otherwise don't take any action. Note that (in my case, at least) the watch list will contain the closest ancestor dir as the one being watched. Make sure not to unwatch a separate path that you actually want to watch (eg. the parent dir)
3. If the dir does exist but is not being watched, `add` it to the watcher.
4. Repeat from step 1. every *n* seconds.

Again, not the best but it definitely works for my use case. Here is a snippet of my hacky solution:

```js
watchDirs = ['/path/to/a/thing/', '/path/to/a/'];

setInterval(() => watchDirs.forEach(async watchDir => {
    
    const getDirFromPath = path => path.replace(/\/+$/, '');
    const getParentDir = dirName => path.dirname(getDirFromPath(dirName));
    const doesDirExist = async dir => {
        try {
            await fs.promises.access(dir);
            return true;
        }
        catch (err) {
            return false;
        }
    };
    
    let dirExists = await doesDirExist(watchDir);
    
    const watchedPaths = watcher.getWatched();
    const fWatchedPathsArr = Object.keys(watchedPaths).map(p => `${p}/`);
    
    if (!dirExists) {
        // The parent directory (and it's parent if the watch dir didn't exist at the point of watch init)
        // become the watch recipient when a dir stops existing. Check if the parent (and it's parent) exists in the watchedPaths
        // list and remove them if they weren't actually supposed to be added.
        const getClosestLivingParent = async dirTree => {
            for (let i = dirTree.length; i > 0; i--) {
                const dirPath = dirTree.slice(0, i).join('/');
                if (await doesDirExist(dirPath)) {
                    return dirPath;
                }
            }
            return '/'; // not sure if I should return '/' or ''
        };
        
        const dirTree = getDirFromPath(watchDir).split(/\/+/g);
        
        const closestLivingParent = await getClosestLivingParent(dirTree);
        const closestLivingParentParent = getParentDir(closestLivingParent);
        
        const fClosestLivingParent = `${closestLivingParent}/`;
        const fClosestLivingParentParent = `${closestLivingParentParent}/`;
        
        const undesiredWatchPathArr = fWatchedPathsArr.filter(p => (
            p === fClosestLivingParent
            && (watchedPaths[getDirFromPath(fClosestLivingParentParent)] || []).length > 0
        ));
        
        undesiredWatchPathArr
            .filter(p => !watchDirs.includes(p)) // do not unwatch paths that the user actually requested
            .forEach(p => {
                watcher.unwatch( getDirFromPath(p) );
                console.log(`Stopped watching for changed on non-existent (removed?) dir ${watchDir}`);
            });
        
    }
    else {
        // Watch dir exists, verify it's in the watch list; otherwise start watching
        const watchDirCheck = getParentDir(getDirFromPath(watchDir));
        if (watchDirCheck in watchedPaths && (watchedPaths[watchDirCheck] || []).length > 0) {
            // This dir exists and it is being watched - we're OK
        }
        else {
            // This dir exists but it's not being watched - start watching
            watcher.add(watchDir);
            console.log(`Re-added dir ${watchDir} to FS watch listener`);
        }
    }
    
}), 10000);
```

This can, of course, be cleaned up to be made more concise.

**Comment by paulmillr** [comment]
Provide a test repository, it's all working fine for me.

**Comment by ezze** [comment]
I faced the similar issue after upgrading to `chokidar@3`. Trying to narrow down the problem I finally stated that it happens when watched directory is a symbolic link.

[Here](https://github.com/ezze/chokidar-issue-917) is a repo to reproduce. My environment:

- Chokidar 3.3.1 (but can be reproduced with >= 3.0.0)
- Node 10.17.0
- Kubuntu 18.04

The issue can't be reproduced with `chokidar@2`.

**Comment by fvictorio** [comment]
The same thing is happening to me but the scenario is simpler (no symlink involved):

- Create `a/b.txt`
- Watch all events on it
- Delete `a/`
- `mkdir a && a/b.txt`

No events are emitted.

**Comment by ComfortablyDumb** [comment]
I have also the same problem with chokidar@2 on Ubuntu and Windows. I hope that the probem will be solved soon.

**Comment by CuddlySheep** [comment]
This issue has already been fixed with Issue #1042.

**Comment by vladiulianbogdan** [comment]
I am seeing this error on Windows. On macOS, it works.
